File: m/rls/rlb/dfexp.c
/* file -  dfexp - expand definition */
#include "m:\rid\rider.h"
#include "m:\rid\dfdef.h"
#include "m:\rid\stdef.h"
#include "m:\rid\chdef.h"
#include "m:\rid\ctdef.h"
#include "m:\rid\fidef.h"
/* code -  messages */
#define E_Ovr  "E-Definition buffer overflow [%s]"
#define E_Mis  "E-Not enough arguments [%s]"
#define E_Xcs  "E-Too many arguments [%s]"
#define E_Syn  "E-Invalid definition operator in [%s]"
static int df_arg (dfTctx *,int ,char *,char **);
/* code -  df_arg - isolates nth argument */
int df_arg(
dfTctx *ctx ,
int nth ,
char *str ,
char *(*res ))
{ char *bal ;
  int sep = ctx->Vsep;
  if( nth < 0)return 0;
  for(;;)  {
    while ( ct_spc (*str)) {++str ;}
    if( *str == 0)return 0;
    *res = str;
    while ( *str != 0) {
      bal = ctx->Pbal;
      while ( *bal != 0) {
        if ( *str == *bal++) {
           str = st_bal (str);break; } }
      if( *str == sep || !*str)break;
      ++str;
    } 
    if( nth-- == 0)break;
    if ( *str != 0) {++str ;}
  } 
  return ( str-*res);
} 
/* code -  df_exp - expand definition */
#define dfARG  0
#define dfREQ  1
#define dfFRM  2
#define dfALL  3
#define dfDEF  4
#define dfTRN  5
int df_exp(
dfTctx *ctx ,
dfTdef *def ,
char *rem ,
char *dst ,
int max ,
int pre ,
int syn )
{ char *bod = def->Pbod;
  char *arg ;
  char *msg ;
  int exp = 0;
  int top = 0;
  int len = 2;
  int cha ;
  int cas ;
  int idx ;
  int cnt ;
  char *bas = dst;
  for(;;)  {
    *dst = 0;
    if ( *bod == 0) {
      if( exp || *rem == 0)return 1;
      *dst++ = ' ', ++len, cha = 'a';
      } else {
      if( ++len >= max){ msg=E_Ovr ; break;}
      if( *bod != pre){ *dst++ = *bod++ ; continue;}
      if( *++bod == pre){ *dst++ = *bod++ ; continue;}
      if( (cha = *bod++&255) == 0){ *dst=0 ; return 1;}
      --len; }
    ++exp;
    if ( cha == 'x') {
      if( df_arg (ctx,top,rem,&arg) == 0)continue;
       msg=E_Xcs;break; }
    switch ( cha) {
    case 'd':
      cas = dfDEF;
     break; case 'e':
      continue;
     break; case ' ':
    case 'n':
       *dst++ = '\n';continue;
     break; case 'a':
      cas = dfALL;
     break; case 'f':
      cas = dfFRM;
     break; case 'r':
      cas = dfREQ;
     break; case 't':
      cas = dfTRN;
     break; default: 
      cas = dfARG;
       }
    if ( cas != dfALL) {
      if ( cas != dfARG) {cha = *bod++ & 255 ;}
      if( cha == 0){ msg=E_Syn ; break;}
      if ( ct_dig (cha)) {
        idx = cha - '0';
      } else if ( ct_upr (cha)) {
        idx = (cha - 'a') + 10;
        } else {
         msg=E_Syn;break; }
      if ( idx > top) {top = idx ;}
      cnt = df_arg (ctx, idx-1, rem, &arg); }
    switch ( cas) {
    case dfALL:
      arg = rem, cnt = st_len (rem);
     break; case dfFRM:
      if ( cnt != 0) {cnt = st_len (arg) ;}
     break; case dfTRN:
      fi_loc (arg, arg);
     break; case dfDEF:
      if( cnt == 0)break;
      idx = 1;
      while ( *bod != 0) {
        if( *bod++ != pre)continue;
        if( (cha=*bod++) == 0)break;
        if ( cha == 'd') {++idx ;}
        if ( cha == 'e') {--idx ;}
        if( idx == 0)break;
      } 
       }
    if( cnt ==  0&& cas == dfREQ){ msg=E_Mis ; break;}
    if( (len += cnt) >= max){ msg=E_Ovr ; break;}
    while ( cnt--) {*dst++ = *arg++ ;}
  } 
  df_rep (ctx, msg, def->Pnam);
  return 0;
} 

File: m/rls/dfexp.r
;???;	literal ^! and ^; required in dfmod
file	dfexp - expand definition
include rid:rider
include	rid:dfdef
include	rid:stdef
include	rid:chdef
include	rid:ctdef
include	rid:fidef

;	df_exp (ctx, def, lin, dst, mxLIN, '#', 0) 
;
;	^^		literal ^
;	^0		definition name
;	^1..^9		parameter 1 to 9
; ???	^A..^Z		caps are parameters A (10) to Z
;	^a		all the command
;	^d#..^e		default ^n to "...", nesting permitted
;	^f#		all command from ^n
;	^p#		"T" if present, "F" if missing
;	^n		insert newline
;	^r#		required parameter n
;	^t#		translated parameter n
;	^x		check excess parameters
;	^<space>	insert newline (same as '^n')
;	^<nl>		inhibit implicit parameters
;
;	^D1
;
;	cmd aaa bbb ccc
;	^1		aaa
;	^2		bbb
; !!!	^a		aaa bbb ccc
;	^d1yyy^e	aaa
;	^d8zzz^e	zzz
;	^f1		aaa bbb ccc
;	^f2		bbb ccc
;	^r1		aaa
;	^r4		%SHE-E-Not enough arguments
;
;	Choice of operators is limited to those that have no other meaning.
;	(*, :, .) must be available for filenames.

	_dfOVR	:= "E-Definition buffer overflow [%s]"
	_dfMIS	:= "E-Not enough arguments [%s]"
	_dfXCS	:= "E-Too many arguments [%s]"
	_dfARG	:= "E-Invalid argument in [%s]"
	_dfSYN	:= "E-Invalid definition operator in [%s]"

	dfARG	:= 0
	dfREQ	:= 1
	dfFRM	:= 2
	dfALL	:= 3
	dfDEF	:= 4
	dfTRN   := 5
	dfPRS   := 6

	df_arg	: (*dfTctx, int, *char, **char) int-	; get nth argument
code	df_trn - translate definition

;	Parse leading identifier
;	Lookup definition
;	Parse definition with line remainder

  func	df_trn
	ctx : * dfTctx
	src : * char
	dst : * char
	len : int
  is	def : * dfTdef 
	idt : [34] char
	ptr : * char = src
	cnt : int = 0
	while *ptr && *ptr gt ' '
	   if cnt lt 32
	   .. idt[cnt++] = *ptr
	   ++ptr
	end
	idt[cnt] = 0
	ptr = st_skp (ptr)
	def = df_loo (ctx, idt)
	pass fail
	df_exp (ctx, def, ptr, dst, len, '^', 0)
	reply that
  end
code	df_exp - expand definition

  func	df_exp
	ctx : * dfTctx				; definition context
	def : * dfTdef				; the definition
	rem : * char				; remainder of command line
	dst : * char				; destination string
	max : int				; dst max (with zero)
	pre : int				; prefix character
	syn : int				; syntax - unused at present
	()  : int				; fail => overrun
  is	bod : * char = def->Pbod		; definition body
	arg : * char				; argument
	msg : * char				; error message
	exp : int = 0				; was explicit
	top : int = 0				; highest parameter seen
	len : int = 2				; we need two extra
	cha : int				;
	cas : int				;
	idx : int				;
	cnt : int				;
	bas : * char = dst			;
						;
     repeat					;
	*dst = 0				; terminate it
	if *bod eq				; all over
	   fine if exp || *rem eq		; was explicit or no remainder
	   *dst++ = ' ', ++len, cha = 'a'	; force remainder (++exp below)
	else					;
	   quit msg=_dfOVR if ++len ge max	; gone too far
	   if *bod eq '|'			; | is newline
	      fine *dst=0 if !bod[1]		; |<nl> ends it
	   .. next *dst++='\n' if *++bod ne '|'	; || is |
	   if (*bod ne '^') && (*bod ne pre)	; old style
	   .. next *dst++ = *bod++		;
	   next *dst++=*bod++ if *bod++ eq *bod ; ^^ is ^
	   fine *dst=0 if (cha = *bod++&255) eq	; ^<nl> ends it 
	.. --len				; adjust for assumption
	++exp					; we have explicit state
	if cha eq 'x'				; checking excess
	   next if df_arg (ctx,top,rem,&arg) eq	; is the last
	.. quit msg=_dfXCS			; has excess arguments
						;
	case cha 				;
	of 'd'    cas = dfDEF			; ^dn...^e default
	of 'e'	  next				; ^e - end of default
	of ' '					; 
	or 'n'	  next  *dst++ = '\n'		; insert newline
	of 'a'    cas = dfALL			; all command
	of 'f'    cas = dfFRM			; from
	of 'p'	  cas = dfPRS			; argument is present
	of 'r'    cas = dfREQ			; required
	of 't'    cas = dfTRN			; translate required
	of other  cas = dfARG			; default
	end case				;
						;
	if cas ne dfALL				; need parameter address
	   cha = *bod++ & 255 if cas ne dfARG	; skip command code
	   quit msg=_dfARG if cha eq		;
	   if ct_dig (cha)			;
	      idx = cha - '0'			;
	   elif ct_upr (cha)			;
	      idx = (cha - 'a') + 10		;
	   else					;
	   .. quit msg=_dfARG			;
	   top = idx if idx gt top		; maximize top
	.. cnt = df_arg (ctx, idx-1, rem, &arg)	; pick it up
						;
	case cas				; 
	of dfALL  arg = rem, cnt = st_len (rem)	; want whole command 
	of dfFRM  cnt = st_len (arg) if cnt ne	; want command tail
	of dfTRN  fi_loc (arg, arg)		; translate to local form
		  cnt = st_len (arg)		;
	of dfDEF  quit if cnt eq		; need the default
		  idx = 1			; nesting level
		  while *bod ne			; skip the default
		     next  if *bod++ ne pre	; not the prefix
		     quit  if (cha=*bod++) eq	; end of line
		     ++idx if cha eq 'd'	; enclosed default
		     --idx if cha eq 'e'	; down a level
		     quit  if idx eq		; end of it
		  end				;
	of dfPRS  arg = "T" if cnt		; argument is present
		  arg = "F" otherwise		; argument is absent
		  cnt = 1			;
	end case				;
						;
	quit msg=_dfMIS if cnt eq && cas eq dfREQ; required parameter missing
	quit msg=_dfOVR if (len += cnt) ge max	; too much of it
	*dst++ = *arg++ while cnt--		; move in the string
     end					;
	df_rep (ctx, msg, def->Pnam)		; report it
	msg = def->Pbod
	PUT("[")
	PUT("%c", *msg++) while *msg && msg lt bod
	PUT("^%s]\n", msg)			;
	fail					;
  end
code	df_arg - isolates nth argument 

  func	df_arg
	ctx : * dfTctx			; the context
	nth : int			; the argument required
	str : * char			; string to vectorize (and destroy)
	res : ** char			; result pointer
	()  : int			; length of it or fail
  is	bal : * char			;
	sep : int = ctx->Vsep		; space or comma
	fail if nth lt			; negative index
      repeat				; the nth
	++str while ct_spc (*str)	; skip leading spaces
	fail if *str eq			; all done
	*res = str			; save start address
	while *str ne			; got more
	   bal = ctx->Pbal		; balance set
	   while *bal ne		; look for balance member
	      if *str eq *bal++		; balance this one?
	   .. .. quit str = st_bal (str);
	   quit if *str eq sep || !*str	; gotcha or eol (from st_bal)
	   ++str			; try again
	end				;
 	quit if nth-- eq		; found it
	++str if *str ne		; skip terminator
      forever				; look at next
	reply str-*res			;
  end
end file
